design/external-authorization-design.md

```diff
@@ -0,0 +1,670 @@
+This type should be shared with other Contour API status types.
+
+This document does not yet define how Contour monitors the support service.
+However, the presence of the `Conditions` field allows Contour to expose the status of the support service fairly directly.
+At minimum, Contour should observe the underlying Kubernetes Service and expose whether there are sufficient healthy Endpoints.
+This information, analogous to Deployments observing and exposing Pod status, can be published as a `Ready` condition.
+
+Each `SupportService` CRD generates a unique upstream Envoy [Cluster][4], which will emit standard Envoy metrics.
+Note that the Envoy cluster name can be non-obvious, so exposing it in status may be helpful.
+
+If the `Service` refers to a Kubernetes `ExternalName`, Contour should program Envoy to send the traffic to the external destination.
+
+The `SupportService` CRD re-uses the `Service` type from the `projectcontour.io/v1` API.
+However, the setting following fields can generate a validation errors:
+
+- `Mirror` must not be set
+- `Weight` must not be set
+- `Protocol` may only be set to `h2`
+
+Note that the `Service` type does not include a field for a namespace name.
+This constrains `SupportService` resources to be in the same namespace as the `Service` they expose,
+ensuring that whoever creates the `SupportService` also has authority over the `Service` backing it.
+
+### HTTPProxy Changes
+
+```Go
+// SupportServiceReference names a SupportService resource.
+type SupportServiceReference struct {
+	// API version of the referent.
+	// +required
+	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,5,opt,name=apiVersion"`
+	// Kind of the referent.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	// +required
+	Kind string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
+	// Namespace of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
+	// Name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// +required
+	Name string `json:"name,omitempty" protobuf:"bytes,3,opt,name=name"`
+}
+
+type AuthorizationPolicy struct {
+	// When true, this field disables client request authentication
+	// for the scope of the policy.
+	//
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+
+	// Context is a set of key/value pairs that are sent to the
+	// authentication server in the check request. If a context
+	// is provided at an enclosing scope, the entries are merged
+	// such that the inner scope overrides matching keys from the
+	// outer scope.
+	//
+	// +optional
+	Context map[string]string `json:"context,omitempty"`
+}
+
+type AuthorizationServer struct {
```

Comment:
It feels like these settings (AuthPolicy, Timeout, Failopen) should be part of the `SupportService` CRD. That's where the contract to the upstream service is defined / configured. 

If many vhosts utilize the same auth server, you'd have to configure these for each vhost and keep in sync.

Reply:
It can't be part of the CRD. Imagine you have 50 services and 1 auth service. You want to gradually migrate the 50 services over to the centralized auth. This means that you move the services over a few at a time, flipping the fail open policy as you go. This means that these parameters, and the others in this struct are specific to the client.

Reply:
I think that use-case is only relevant once. The backend auth I doubt would have different timeouts per vhost. If that was the case you'd have different CRDs to represent the different backends. But given many use the same backend auth, you now need to sync those settings manually across many vhosts.

Reply:
Yeh, arguably there is a use case for timeout being on the CRD.